yamux/src/connection/stream/flow_control.rs

```diff
@@ -0,0 +1,305 @@
+use std::{cmp, sync::Arc, time::Instant};
+
+use parking_lot::Mutex;
+
+use crate::{connection::rtt::Rtt, Config, DEFAULT_CREDIT};
+
+#[derive(Debug)]
+pub(crate) struct FlowController {
+    config: Arc<Config>,
+    last_window_update: Instant,
+    /// See [`Connection::rtt`].
+    rtt: Rtt,
+    /// See [`Connection::accumulated_max_stream_windows`].
+    accumulated_max_stream_windows: Arc<Mutex<usize>>,
```

Comment:
Could this live outside of `FlowController` instead and you pass a mutable reference into the functions where you need it?

It feels odd that an implementation detail about shared, mutable state leaks into the implementation of this algorithm.

Reply:
Same thing with `Arc<Config>` and `Rtt` actually. It would be nice if we could express this algorithm in a more functional way and pass these dependencies in whenever needed. Thoughts?